### PR TITLE
fix(vector sink): retry on failed connection

### LINKED ISSUE #7624: HTTP test server needs to provide gRPC headers to indicate response status
In order to test end to end acknowledgement properly the HTTP test server needs to be able to provide the required gRPC headers to indicate the response status. Currently the header is missing and we get this error:

```
message: "grpc-status header missing, mapped from HTTP status code 200"
```

There is some useful information about the protocol [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).

### LINKED ISSUE #8648: vector v2 sink does not retry on connection refused
When the vector v2 sink is pointed at an address that refuses connections, it tries once and then immediately marks the batch as having an error. A refused connection is a potentially transient condition that should instead cause a retry. While the batch status indicates a transient error for end-to-end acknowledgements, when the source connected to this sink cannot retry events (ie Kafka), this results in lost events.

It is not clear if the v2 prep work #8401 changes this behavior.

Ref https://github.com/timberio/vector/discussions/8563#discussioncomment-1154983

### PR DESCRIPTION
**Pull request by JeanMertz**
fixes #8648.
fixes #7624.

### CODE REVIEW

**Comment by JeanMertz** [comment]
> Should we instead always treat `Unknown` errors as retryable?

Actually, what I'm more curious about is if we have any guidelines on what types of error _should_ fail the component (and shut down Vector)?

For example, currently we also fail on internal server errors, or permission errors, but it could be reasoned that internal server errors are often temporarily, and permission errors _might_ be due to invalid credentials, but even so that can still be resolved on the server (e.g. by matching the credentials used by the client).

So really what I'm asking is, which types of errors should _never_ be retried, if any?

**Comment by spencergilbert** [comment]
> > Should we instead always treat `Unknown` errors as retryable?
> 
> Actually, what I'm more curious about is if we have any guidelines on what types of error _should_ fail the component (and shut down Vector)?
> 
> For example, currently we also fail on internal server errors, or permission errors, but it could be reasoned that internal server errors are often temporarily, and permission errors _might_ be due to invalid credentials, but even so that can still be resolved on the server (e.g. by matching the credentials used by the client).
> 
> So really what I'm asking is, which types of errors should _never_ be retried, if any?

My gut instinct would be fail on permission errors, and retry (with backoff?) for internal server errors... that said I don't have any good reason why they should be different...

**Comment by jszwedko** [comment]
Yeah, I'm inclined to agree with @spencergilbert that we shouldn't retry the equivalent of 400-level HTTP errors (that is, bad requests) which'll never succeed, but retry anything else. I'm happy to see us go with whatever the current `http` sink does though until we can revisit this more holistically.

I made a note on https://github.com/timberio/vector/issues/8728 to cover this topic there.

**Comment by spencergilbert** [comment]
> Yeah, I'm inclined to agree with @spencergilbert that we shouldn't retry the equivalent of 400-level HTTP errors (that is, bad requests) which'll never succeed, but retry anything else. I'm happy to see us go with whatever the current `http` sink does though until we can revisit this more holistically.
> 
> I made a note on #8728 to cover this topic there.

Alternatively, I think the fact that we own both ends of this component we're free to handle however we see fit - which I think @JeanMertz mentioned during standup

File: src/sinks/vector/v2.rs [hunk 1]
```diff
@@ -263,11 +263,14 @@ impl RetryLogic for VectorGrpcRetryLogic {
...
+        use tonic::Code::*;
+
         match err {
-            Error::Request { source } => !matches!(
-                source.code(),
-                tonic::Code::Unknown | tonic::Code::Internal | tonic::Code::PermissionDenied
-            ),
+            Error::Request { source } => match source.code() {
+                Unknown if source.message().contains("tcp connect error") => true,
+                Unknown | Internal | PermissionDenied => false,
```
Comment (bruceg) [inline src/sinks/vector/v2.rs:271]: I think `Internal` errors should definitely be retriable here, as they are reporting a potentially transient condition on the server. I _think_ `Unknown` should also be retriable, but it depends somewhat what all errors are grouped into this category.

**Review by bruceg — CHANGES_REQUESTED** [review]

**Comment by JeanMertz** [comment]
This one should be good to go. @bruceg feel like giving it another look?

**Review by bruceg — APPROVED** [review]
LGTM

**Pull request merged**